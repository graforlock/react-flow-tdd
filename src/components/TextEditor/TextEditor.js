// @flow
import React from 'react'
import { connect } from 'react-redux'

import { fetchUpdateIdea, setEditorBody, setEditorTitle, clearEditor } from '../../actions'
import type { ChildrenArray } from 'react'
import type { EditingIdea, EditingField } from '../../reducers/editorReducer'
import type { IdeaRequest } from '../../services/ideaService'
import type { IdeasFetchAction } from '../../actions/actionTypes'

type State = {
  editing: boolean,
  letterCount: ?number
}

type Props = {
  children: ChildrenArray<any>[],
  idea: IdeaRequest,
  field: EditingField,
  editedIdea: ?EditingIdea,
  fetchUpdateIdea: (idea: IdeaRequest) => IdeasFetchAction,
  setEditorBody: Function,
  setEditorTitle: Function,
  clearEditor: Function
}

const MAX_LENGTH: number = 75

class TextEditor extends React.Component<Props, State> {
  state: State = {
    editing: false,
    letterCount: null
  }

  handleInput = e => {
    const newCount: number = MAX_LENGTH - e.target.value.length
    this.setState({ letterCount: newCount })
  }

  switchMode = () => {
    const { field, idea, setEditorBody, setEditorTitle } = this.props
    if (field === 'TITLE') setEditorTitle(idea.id)
    if (field === 'BODY') setEditorBody(idea.id)
    // const { editing } = this.state
    // this.setState({ editing: !editing })
  }

  handleBlur = (e) => {
    const { clearEditor, idea, field } = this.props
    const { target: { value } } = e
    if (field === 'TITLE') {
      this.props.fetchUpdateIdea({
        ...idea, ...{ title: value }
      })
    } else if (field === 'BODY') {
      this.props.fetchUpdateIdea({
        ...idea, ...{ body: value }
      })
    }
    clearEditor()
  }

  renderChildren (): ChildrenArray<any>[] {
    const { children } = this.props
    return React.Children.map(children, (element) =>
        React.cloneElement(element, { onClick: this.switchMode }))
  }

  render () {
    const { editedIdea, idea, field } = this.props
    const { letterCount } = this.state
    const displayCount = letterCount && letterCount <= 15
    return (
      <section className='texteditor__container'>
        { editedIdea && editedIdea.id === idea.id && editedIdea.type === field
          ? <textarea className='idea__textarea' maxLength={MAX_LENGTH}
            autoFocus defaultValue={idea[field.toLowerCase()]}
            onBlur={e => this.handleBlur(e)}
            onInput={e => this.handleInput(e)}
          />
          : this.renderChildren() }
        <div>{ displayCount && letterCount }</div>
      </section>
    )
  }
}

const mapStateToProps = ({ editor: { editedIdea } }) => {
  return { editedIdea }
}

export default connect(mapStateToProps, {
  setEditorBody,
  setEditorTitle,
  clearEditor,
  fetchUpdateIdea
})(TextEditor)
