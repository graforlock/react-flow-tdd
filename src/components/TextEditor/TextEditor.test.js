// @flow
import React from 'react'
import type { ShallowWrapper } from 'enzyme'
import { shallow } from 'enzyme'

import { TextEditor } from './TextEditor'

describe('TextEditor', () => {
  let wrapper: ShallowWrapper
  const idea = {
    id: 1,
    title: 'Title',
    body: '... body ...',
    created_date: new Date('2049')
  }

  beforeEach(() => {
    wrapper = shallow(
      <TextEditor clearEditor={jest.fn()}
        field='TITLE' idea={idea}
        fetchUpdateIdea={jest.fn()}
        setEditorTitle={jest.fn()}
        setEditorBody={jest.fn()}
        editedIdea={jest.fn()} children={[<div />]} />
      )
  })

  it('renders component and matches current default snapshot', () => {
    expect(wrapper).toMatchSnapshot()
  })
})
