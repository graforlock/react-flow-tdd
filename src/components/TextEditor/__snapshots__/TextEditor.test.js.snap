// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextEditor renders component and matches current default snapshot 1`] = `
<section
  className="texteditor__container"
>
  <div
    key=".0"
    onClick={[Function]}
  />
  <div />
</section>
`;
